Fix #2269: Send emails to all qualified users of query.

prasanna08 opened the pull request:
This PR adds functionality to send email to qualified users of a query (milestone 6).

- Added email dashboard result page where query submitter can give email body, subject and other options.
- Added backend handlers for email dashboard result.
- Added tests to make sure everything is consistent.

Thanks.

prasanna08 commented:
Hi @seanlip, added milestone 6 of #2269. Also should we display link to email dashboard result page of a query in past queries table on email dashboard page ?

Thanks

seanlip commented:
Thanks! I'll take a look later, but in response to your question, my first impression is yes. (Why wouldn't we?)

seanlip commented on core/controllers/email_dashboard_test.py:
```diff
@@ -117,3 +118,300 @@ def test_that_exception_is_raised_for_invalid_input(self):
+
+            # Check that emails are sent to max n qualified users.
+            # One email is sent to submitter for query completion and second
+            # is sent to one of the 2 qualified users.
```
Can you make sure that this is the case by checking the recipient ids?

seanlip commented on core/controllers/email_dashboard_test.py:
```diff
@@ -117,3 +118,300 @@ def test_that_exception_is_raised_for_invalid_input(self):
+                        'email_subject': 'subject',
+                        'email_body': 'body',
+                        'email_option': 'cancel',
+                        'num_of_users_to_send': None,
```
Erh -- this is ugly. I think you should have separate handlers in the controller for cancelling and sending.

prasanna08 replied:
I don't understand what sort of logic you are suggesting ?
Do you suggest that we should have a separate handler (determining which handler to call in front end) for sending and cancelling and in sending handler we check what is value of `max_recipient`, if it is None then send to all or else send to first max_recipient users ?
That way, intermediary function in `user_query_services.py` will not be required. Each handler will send directly by passing appropriate recipient_ids to email_manager or cancel query directly.

seanlip replied:
Yes, that's exactly right (re the second paragraph).

You might still want separate intermediate domain-layer functions for the logic -- the controllers shouldn't be doing much stuff, especially with regards to querying the datastore.

seanlip commented on core/domain/user_query_services.py:
```diff
@@ -38,3 +40,30 @@ def save_new_query_model(
+
+
+def send_email_to_qualified_users(
+        query_id, email_subject, email_body, email_intent, email_option,
```
This needs a docstring explaining what the args are. Please follow the style in e.g. core/domain/feedback_services.py.

Also, please use separate functions for the different email_option cases. Using the same function to handle multiple completely different modes is an antipattern.

seanlip commented on core/domain/user_query_services.py:
```diff
@@ -38,3 +40,30 @@ def save_new_query_model(
+        query_model.sent_email_model_id = bulk_email_model_id
+        query_model.put()
+    elif email_option == feconf.USER_QUERY_EMAIL_OPTION_CUSTOM:
+        recipient_ids = query_model.user_ids[:num_of_users_to_send]
```
Btw, what's the max number of users that we can pack into user_ids before we hit the GAE memory limit? We should add a warning about that if this can plausibly occur.

prasanna08 replied:
Seems like we have an issue here. According to [this](http://stackoverflow.com/questions/15377119/gae-ndb-design-performance-and-use-of-repeated-properties) we should not store more than 100-1000 values in repeated properties (my bad, sorry for not cross checking this earlier).  We will have more than 1000 values usually. So we should use JsonProperty (probably with compressed=True). We can do this change in UserQueryModel easily but this will affect BulkEmailModel where we can query on each individual user (which will not be possible in JsonProperty). We have to think of some new way to store emails sent to individual users and then change both models. What do you think  ?

seanlip replied:
Hm, yeah -- this is tricky. We do have a SentEmailModel -- what if we do a put_multi() into that? But that's going to be expensive for large recipient lists. That said, I don't think we'll be sending bulk emails that often...

prasanna08 replied:
So you suggest we put multiple in models in SentEmailModel ? I still think it will cost a significant amount of storage space.

prasanna08 replied:
Yup second idea is batter. Should I implement it in current PR ?

seanlip commented on core/domain/user_query_services.py:
```diff
@@ -38,3 +40,30 @@ def save_new_query_model(
+    query_model = user_models.UserQueryModel.get(query_id)
+    query_model.query_status = feconf.USER_QUERY_STATUS_ARCHIVED
+
+    if email_option == feconf.USER_QUERY_EMAIL_OPTION_ALL:
```
Perhaps you can signal this case by having num_of_users_to_send be None. (Unless None already indicates something else, like "not yet populated"?)

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -354,6 +354,8 @@ class UserQueryModel(base_models.BaseModel):
     user_ids = ndb.StringProperty(indexed=True, repeated=True)
     # Id of the user who submitted the query.
     submitter_id = ndb.StringProperty(indexed=True, required=True)
+    # Id of the email model having information of sent emails.
```
Be specific: give the actual class name of the model. E.g.

Id of the SentEmailModel(?) instance containing information about sent emails.

Also, I suggest Id --> ID (here and above).

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboard.js:
```diff
@@ -75,6 +75,11 @@ oppia.controller('EmailDashboard', [
       });
     };
+
+    $scope.showLink = function(submitter, status) {
```
showLinkToResultPage

seanlip commented on core/controllers/email_dashboard.py:
```diff
@@ -157,3 +161,40 @@ def get(self):
+        email_subject = data['email_subject']
+        email_body = data['email_body']
+        email_option = data['email_option']
+        num_of_users_to_send = data['num_of_users_to_send']
```
Can you change this, here and elsewhere, to max_recipients?

seanlip commented on core/controllers/email_dashboard.py:
```diff
@@ -157,3 +161,40 @@ def get(self):
         self.render_json(data)
+
+
+class EmailDashboardResult(base.BaseHandler):
```
EmailDashboardResultPage

seanlip commented on core/controllers/email_dashboard.py:
```diff
@@ -157,3 +161,40 @@ def get(self):
+            raise self.InvalidInputException('400 Invalid query id.')
+
+        self.values.update({
+            'query_id': query_id,
```
We are actually trying to get rid of Jinja. I suggest doing the usual thing of defining a corresponding EmailDashboardResultHandler class with get() and post() methods, similar to what we do elsewhere. The get() method is called on controller init and supplies the query id. Then you don't need to use GLOBALS, too.

prasanna08 replied:
I don't know how we can get `query_id` this way. `query_id` is embedded in URL so only handler for that URL has access to it. Even if we do send request to some other handler, how that handler will be able to identify which query's id we want ?
It can be extracted from URL directly, if Jinja way is not possible.

seanlip replied:
Ah, I see, my bad. Yes, I think you should extract it directly from the URL.

seanlip commented on feconf.py:
```diff
@@ -554,7 +554,10 @@ def get_empty_ratings():
 USER_QUERY_STATUS_ARCHIVED = 'archived'
 USER_QUERY_STATUS_FAILED = 'failed'
-
+# Possible email option for sending emails using query.
```
We should get rid of these and have separate handlers instead.

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard_result.html:
```diff
@@ -0,0 +1,73 @@
+
+{% block footer_js %}
+  {{ super() }}
+  <script type="text/javascript">
```
This should be removed.

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardResult.js:
```diff
@@ -0,0 +1,91 @@
+    var EMAIL_POST_HANDLER = '/emaildashboardresult';
+
+    $scope.submitEmail = function() {
+      var postLink = EMAIL_POST_HANDLER + '/' + $scope.queryId;
```
resultHandlerUrl

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardResult.js:
```diff
@@ -0,0 +1,91 @@
+oppia.controller('EmailDashboardResult', [
+  '$scope', '$http', '$window', '$timeout',
+  function($scope, $http, $window, $timeout) {
+    var EMAIL_POST_HANDLER = '/emaildashboardresult';
```
RESULT_HANDLER_URL_PREFIX

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardResult.js:
```diff
@@ -0,0 +1,91 @@
+          $timeout(function() {
+            $window.location.href = EMAIL_DASHBOARD_PAGE;
+          }, 4000);
+        });
```
Should do something on .error()

prasanna08 replied:
We can display a message saying please try again later. Do you have something else in mind ?

seanlip replied:
Just display a message saying "There was an error processing your request. Please report this bug." or similar, I think.

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardResult.js:
```diff
@@ -0,0 +1,91 @@
+          data: data
+        }).success(function() {
+          $scope.emailSubmitted = true;
+          $timeout(function() {
```
I think all the fields should be grayed out and buttons disabled while a submit is in progress. You can maintain a $scope.submitIsInProgress boolean variable for this.

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardResult.js:
```diff
@@ -0,0 +1,91 @@
+      }
+      if ($scope.email_option === 'custom' &&
+        $scope.num_of_users_to_send === null) {
+        $scope.invalid.num_users = true;
```
Try and keep the names consistent -- don't use both `num_users` and `num_of_users_to_send`. (Although I suggested changing this to `max_recipients` -- see above.)

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardResult.js:
```diff
@@ -0,0 +1,91 @@
+      }
+    };
+
+    $scope.resetEmail = function() {
```
resetForm (you're resetting the form, not the email)

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardResult.js:
```diff
@@ -0,0 +1,91 @@
+
+    $scope.queryId = GLOBALS.queryId;
+    $scope.email_option = 'all';
+    $scope.email_subject = '';
```
$scope.emailSubject, $scope.emailBody, $scope.emailIntent, $scope.maxRecipients, etc.

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardResult.js:
```diff
@@ -0,0 +1,91 @@
+      num_users: false
+    };
+    $scope.num_of_users_to_send = null;
+    $scope.possible_email_intents = [
```
$scope.POSSIBLE_EMAIL_INTENTS (constant)

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard.html:
```diff
@@ -63,10 +63,13 @@
             <td><[q.id]></td>
             <td><[q.created_on]></td>
-            <td><[q.submitter_id]></td>
+            <td><[q.submitter]></td>
```
submitter_username

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard.html:
```diff
@@ -63,10 +63,13 @@
             <td><[q.num_qualified_users]></td>
             <td><button ng-click="recheckStatus($index)">Re-check status</button></td>
+            <td ng-if="showLink(q.submitter, q.status)">
+              <a href="/emaildashboardresult/<[q.id]>"><button>Result</button></a>
```
ng-href

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard_result.html:
```diff
@@ -0,0 +1,73 @@
+
+{% block header_js %}
+  {{ super() }}
+  <style>
```
This isn't JS -- put it in the content block.

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard_result.html:
```diff
@@ -0,0 +1,73 @@
+    .invalid-input {
+      color: red;
+    }
+    .success {
```
Why two `.success` rules?

prasanna08 replied:
Oh, I did not realise I already had one.
Thanks for pointing out.

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard_result.html:
```diff
@@ -0,0 +1,73 @@
+{% block content %}
+  <div class="oppia-email-form" ng-controller="EmailDashboardResult">
+    <p ng-if="emailSubmitted" class="success">Emails sent successfully. You will be redirected to Email Dashboard page.</p>
+    <p ng-if="invalid.subject" class="invalid-input">Please enter valid email subject</p>
```
valid --> a valid; ditto below.

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard_result.html:
```diff
@@ -0,0 +1,73 @@
+    <p ng-if="emailSubmitted" class="success">Emails sent successfully. You will be redirected to Email Dashboard page.</p>
+    <p ng-if="invalid.subject" class="invalid-input">Please enter valid email subject</p>
+    <p ng-if="invalid.body" class="invalid-input">Please enter valid email body</p>
+    <p ng-if="invalid.num_users" class="invalid-input">Please enter valid no. of recipients</p>
```
no. --> number

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard_result.html:
```diff
@@ -0,0 +1,73 @@
+
+{% block content %}
+  <div class="oppia-email-form" ng-controller="EmailDashboardResult">
+    <p ng-if="emailSubmitted" class="success">Emails sent successfully. You will be redirected to Email Dashboard page.</p>
```
You will be redirected to Email Dashboard page. --> Redirecting to the Email Dashboard...

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard_result.html:
```diff
@@ -0,0 +1,73 @@
+    <form class="form" ng-submit="submitEmail()">
+      Email subject: <textarea rows="2" cols="10" placeholder="Enter email subject here..." ng-model="email_subject"></textarea>
+      <br>
+      Email body: <textarea rows="10" cols="20" placeholder="Enter email body here..." ng-model="email_body"></textarea>
```
Only 20 columns? Pretty short email body...

prasanna08 replied:
Well `textarea` is flexible tag. This is just default size but in front end you can narrow it or expand it it from corner of the text area. I put it small so that everything on page is covered without scrollbar.

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard_result.html:
```diff
@@ -0,0 +1,73 @@
+      Email option: <input type="radio" name="email" ng-model="email_option" value="all" checked> Send to all
+      <input type="radio" name="email" ng-model="email_option" value="custom"> Send to max N users:
+      <input type="number" ng-model="num_of_users_to_send" min="0" placeholder="Enter number of users...">
+      <input type="radio" name="email" ng-model="email_option" value="cancel"> Cancel<br><br>
```
This should be a separate button, all on its own.

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard_result.html:
```diff
@@ -0,0 +1,73 @@
+      <input type="radio" name="email" ng-model="email_option" value="custom"> Send to max N users:
+      <input type="number" ng-model="num_of_users_to_send" min="0" placeholder="Enter number of users...">
+      <input type="radio" name="email" ng-model="email_option" value="cancel"> Cancel<br><br>
+      <input type="submit" value="submit email">
```
Capitalize: "Submit Email"? Also maybe this should have class "btn btn-success"...

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard_result.html:
```diff
@@ -0,0 +1,73 @@
+      <input type="number" ng-model="num_of_users_to_send" min="0" placeholder="Enter number of users...">
+      <input type="radio" name="email" ng-model="email_option" value="cancel"> Cancel<br><br>
+      <input type="submit" value="submit email">
+      <input type="reset" value="reset" ng-click="resetEmail()">
```
And this one has class "btn btn-default". Also: "Reset" (capitalize)?

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard_result.html:
```diff
@@ -0,0 +1,73 @@
+      <br>
+      Email intent: <select ng-model="email_intent" ng-options="intent for intent in possible_email_intents"></select><br><br>
+      Email option: <input type="radio" name="email" ng-model="email_option" value="all" checked> Send to all
+      <input type="radio" name="email" ng-model="email_option" value="custom"> Send to max N users:
```
Ah, I see why the backend was the way it was. Don't let the UI force the design of your backend, though -- that's kind of the tail wagging the dog. If you need to do some transformations, do it in the JS here before sending stuff to the backend.

prasanna08 commented:
Hi @seanlip added new commit addressing past comments.

- Changed UserQueryModel and BulkEmailModel as we discussed (use JsonProperty) and also added new UserBulkEmailsModel.
- Added separate controller to handler cancelling query requests.
- Changed logic of all / max_recipients as we discussed (using `max_recipients` value to distinguish 'all' vs 'custom' case).
- Many small frontend and backend changes for variable name changes, error reporting, frontend view changes, etc..

Thanks,
Prasanna

seanlip commented on core/domain/user_query_services.py:
```diff
@@ -38,3 +40,41 @@ def save_new_query_model(
+        email_subject: str. Subject of the email to be sent.
+        email_body: str. Body of the email to be sent.
+        email_intent: str. Intent of the email.
+        max_recipients: integer. Number of maximum recipients to send emails.
```
nit: integer --> int (this is the Python datatype)

Number of maximum --> Maximum number of
send emails. --> send emails to.

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -351,9 +351,12 @@ class UserQueryModel(base_models.BaseModel):
+    user_ids = ndb.JsonProperty(default=[], compressed=True)
+    # ID of the user who submitted the query.
     submitter_id = ndb.StringProperty(indexed=True, required=True)
+    # ID of the instance of BulkEmailModel which stores infromation
```
infromation --> information

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -372,3 +375,11 @@ def fetch_page(cls, page_size, cursor):
+class UserBulkEmailsModel(base_models.BaseModel):
+    """Model to store IDs BulkEmailModel sent to a user.
+
+    Instances of this class are keyed by the user id"""
```
Add full stop at end; move """ to next line.

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -372,3 +375,11 @@ def fetch_page(cls, page_size, cursor):
+    """Model to store IDs BulkEmailModel sent to a user.
+
+    Instances of this class are keyed by the user id"""
+    # IDs of all sent BulkEmailModels.
```
IDs of all BulkEmailModels that correspond to bulk emails sent to this user.

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -372,3 +375,11 @@ def fetch_page(cls, page_size, cursor):
+
+    Instances of this class are keyed by the user id"""
+    # IDs of all sent BulkEmailModels.
+    sent_email_models_ids = ndb.StringProperty(indexed=True, repeated=True)
```
sent_email_model_ids

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardResult.js:
```diff
@@ -0,0 +1,117 @@
+        var data = {
+          email_subject: $scope.emailSubject,
+          email_body: $scope.emailBody,
+          email_intent: $scope.emailIntent
```
Maybe simpler to add, after this line:

    max_recipients: ($scope.emailOption !== 'all' ? $scope.maxRecipients : null)

seanlip commented on core/templates/dev/head/pages/email_dashboard/email_dashboard_result.html:
```diff
@@ -0,0 +1,73 @@
+      Email intent: <select ng-model="emailIntent" ng-options="intent for intent in POSSIBLE_EMAIL_INTENTS"></select><br><br>
+      Email option: <input type="radio" name="email" ng-model="emailOption" value="all" checked> Send to all
+      <input type="radio" name="email" ng-model="emailOption" value="custom"> Send to max N users:
+      <input type="number" ng-model="maxRecipients" min="0" placeholder="Enter number of users..."><br><br>
```
Perhaps you could disable this field if emailOption is set to 'all'.

seanlip commented:
Hm, I noticed a strange failure in the e2e tests. I think it's because the errors that show when a YouTube video is embedded have changed.

Would you mind fixing this? I think the change is straightforward -- it would be to the error message in checkForConsoleErrors() at the end of core/tests/protractor/richTextComponents.js.

Thanks!

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboardResult.js:
```diff
@@ -50,7 +50,9 @@ oppia.controller('EmailDashboardResult', [
+          max_recipients: (
+            $scope.emailOption !== 'all' ? $scope.max_recipients : null)
         };
         if ($scope.emailOption === 'all') {
```
But now you don't need these lines.

prasanna08 replied:
Yup my bad.

seanlip commented:
@prasanna08 -- I think one comment was incompletely applied. Otherwise, LGTM!

seanlip commented:
LGTM
